include/interface/blas1_interface_sycl.hpp

```diff
@@ -338,67 +338,108 @@ void _asum(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
 }
 
 /**
- * \brief ICAMAX finds the index of the first element having maximum
+ * \brief IAMAX finds the index of the first element having maximum
  * @param _vx  VectorView
  * @param _incx Increment in X axis
  */
-template <typename ExecutorType, typename T, typename ContainerT>
-size_t _iamax(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
-              int _incx) {
-  auto my_vx = vector_view<T, ContainerT>(_vx, _vx.getDisp(), _incx, _N);
+template <typename ExecutorType, typename T, typename ContainerT, typename I,
+          typename ContainerI>
+void _iamax(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
+            int _incx, vector_view<I, ContainerI> _rs) {
+  vector_view<T, ContainerT> my_vx(_vx, _vx.getDisp(), _incx, _N);
+  vector_view<I, ContainerI> my_rs(_rs, _rs.getDisp(), 1, 1);
 #ifdef VERBOSE
   my_vx.printH("VX");
 #endif  //  VERBOSE
-
-  auto localSize = 256;
-  auto nWG = 512;
-  std::vector<IndVal<T>> valT1(nWG);
-  auto val1 = vector_view<IndVal<T>, std::vector<IndVal<T>>>(valT1, 0, 1, nWG);
+  size_t localSize = 256, nWG = 512;
+  auto tupOp = TupleOp<vector_view<T, ContainerT>>(my_vx);
+  std::vector<IndVal<T>> valT1(nWG,
+                               IndVal<T>(std::numeric_limits<size_t>::max(),
+                                         std::numeric_limits<T>::min()));
+  cl::sycl::buffer<IndVal<T>, 1> bvalT1(valT1.data(), cl::sycl::range<1>{nWG});
+  BufferVectorView<IndVal<T>> val1(bvalT1, 0, 1, nWG);
+  /* vector_view<IndVal<T>,BufferVectorView<IndVal<T>>> val1(valT1, 0, 1, nWG);
+   */
   auto assignOp1 =
-      make_maxIndReducAssignNewOp3(val1, my_vx, localSize, nWG * localSize);
-  ex.execute(assignOp1);
-
-  std::vector<IndVal<T>> valT2(1);
-  auto val2 = vector_view<IndVal<T>, std::vector<IndVal<T>>>(valT1, 0, 1, 1);
-  auto assignOp2 = make_maxIndReducAssignNewOp2(val2, val1, localSize, nWG);
-  ex.execute(assignOp2);
+      make_maxIndReducAssignNewOp2(val1, tupOp, localSize, localSize * nWG);
+  ex.reduce(assignOp1);
+  auto assignOp2 = make_maxIndReducAssignNewOp2(my_rs, val1, localSize, nWG);
+  ex.reduce(assignOp2);
 #ifdef VERBOSE
   std::cout << "ind = " << val1.eval(0).getInd() << std::endl;
 #endif  //  VERBOSE
-  return val1.eval(0).getInd();
```

Comment:
So, there is a version of _iamax which returns nothing, but writes the index-value pair into the _rs parameter, and there is another function which uses it and returns size_t.